utils/s3/client.cc

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/util/short_streams.hh>
+#include <seastar/http/request.hh>
+#include "utils/s3/client.hh"
+#include "utils/memory_data_sink.hh"
+#include "utils/chunked_vector.hh"
+#include "log.hh"
+
+namespace s3 {
+
+static logging::logger s3l("s3");
+
+future<> ignore_reply(const http::reply& rep, input_stream<char>&& in_) {
+    auto in = std::move(in_);
+    co_await util::skip_entire_stream(in);
+}
+
+client::client(socket_address addr, private_tag)
+        : _addr(std::move(addr))
+        , _host(to_sstring(_addr))
+        , _http(_addr)
+{
+}
+
+shared_ptr<client> client::make(socket_address addr) {
+    return seastar::make_shared<client>(std::move(addr), private_tag{});
+}
+
+future<uint64_t> client::get_object_size(sstring object_name) {
+    s3l.trace("HEAD {}", object_name);
+    auto req = http::request::make("HEAD", _host, object_name);
+    uint64_t len = 0;
+    co_await _http.make_request(std::move(req), [&len] (const http::reply& rep, input_stream<char>&& in_) mutable -> future<> {
+        len = rep.content_length;
+        return make_ready_future<>(); // it's HEAD with no body
+    });
+    co_return len;
+}
+
+future<temporary_buffer<char>> client::get_object_contiguous(sstring object_name, std::optional<range> range) {
+    s3l.trace("GET {} contiguous", object_name);
+    auto req = http::request::make("GET", _host, object_name);
+    http::reply::status_type expected = http::reply::status_type::ok;
+    if (range) {
+        req._headers["Range"] = format("bytes={}-{}", range->off, range->off + range->len - 1);
```

Comment:
Worth dumping the byte range in the trace log, so to help with debugging. For example, HTTP server might ignore the range header if there's anything wrong with it and return 200 response type on completion.

Reply:
> Worth dumping the byte range in the trace log

You mean the header itself, don't you? Not the actual bytes from the body :sweat_smile: